File: frontend/projects/irrtirer/src/app/shared/mosaic-generator/orderer/section-layout/section-layout-preparer.ts
import { IterableUnion } from '../../../../core/helpers/iterable-union';
import { PointToPolygonDistanceCalculator } from '../../../../core/helpers/polygon/point-to-polygon-distance-calculator';
import { PresenceInPoligonHelper } from '../../../../core/helpers/polygon/presence-in-polygon-helper';
import { RandomHelper } from '../../../../core/helpers/random-helper';
import { Black } from '../../../../core/models/color.model';
import { Vector } from '../../../../core/models/vector.model';
import { BluredImageObject } from '../../color-compatibility/blured-image-object.model';
import { GeometryObject } from '../../models/geometry-object.model';
import { Tile } from '../../models/tile';
import { TileTransform } from '../../models/tile-transform.model';
import { PopulationParams } from '../../sectors/generation-params/population-params.interface';
import { SectionModel } from '../../sectors/section.model';
import { SectionPointFilteredTray } from '../../tray/section-point-filtered-tray';
import { PossibleLayout } from './possible-layout';
import { PossibleLayoutEvaluator } from './possible-layout-evaluator';

export class SectionLayoutPreparer {
    private readonly neighbourTiles: TileTransform[];
    private readonly populationParams: PopulationParams;
    private readonly evaluator: PossibleLayoutEvaluator;

    constructor(private pixelSource: BluredImageObject, private section: SectionModel, neighbourTiles: TileTransform[]) {
        this.neighbourTiles = neighbourTiles;
        this.populationParams = section.parent.populationParams;
        this.evaluator = new PossibleLayoutEvaluator(section.parent.evaluationParams, pixelSource, section);
    }

    public createInitialLayout(avalibleTiles: Tile[]): PossibleLayout {
        const layout: PossibleLayout = new PossibleLayout(this.section);
        layout.usedTilesIndexes = new Set<number>();
        layout.tilesAvalibleForLayout = avalibleTiles;

        const pointFilteredTray: SectionPointFilteredTray = new SectionPointFilteredTray(avalibleTiles, layout.usedTilesIndexes);

        for (let i = 0; i < this.populationParams.countOfRandomingTrianglePosition && pointFilteredTray.hasAnyAvailableTiles(); ++i) {
            this.TryAddTile(layout, pointFilteredTray);
        }

        layout.rate = this.evaluator.evaluate(layout);
        return layout;
    }

    public ModifyLayout(layout: PossibleLayout, count: number): PossibleLayout {
        throw new Error();
    }

    private TryAddTile(layout: PossibleLayout, pointFilteredTray: SectionPointFilteredTray): boolean {
        const randomPoint: Vector = RandomHelper.randomPointInTriangle(this.section.triangle.vertices);
        if (!this.testIfPointIsInsideTile(new IterableUnion([this.neighbourTiles, layout.getTilesLayout()]), randomPoint)) {
            pointFilteredTray.setFilterForPoint({
                maxInnerRadius:
                    this.getMinDistanceToNeighbourTiles(new IterableUnion([this.neighbourTiles, layout.getTilesLayout()]), randomPoint) ??
                    Number.MAX_VALUE,
                preferredColor: this.pixelSource.getPictureColorAtPosition(randomPoint) ?? Black,
                countOfColorMatchingAttempts: this.populationParams.countOfColorMatchingAttempts,
            });

            const tileTransform: TileTransform = this.tryFindTileTransformForPoint(pointFilteredTray, layout.getTilesLayout(), randomPoint);
            if (tileTransform != null) {
                this.addTile(layout, pointFilteredTray, tileTransform);
                return true;
            }
        }

        return false;
    }

    private TryMoveTile(layout: PossibleLayout, pointFilteredTray: SectionPointFilteredTray): boolean {
        throw new Error();
    }

    private TryRotateTile(layout: PossibleLayout, pointFilteredTray: SectionPointFilteredTray): boolean {
        throw new Error();
    }

    private TryReplaceTile(layout: PossibleLayout, pointFilteredTray: SectionPointFilteredTray): boolean {
        throw new Error();
    }

    private tryFindTileTransformForPoint(
        pointFilteredTray: SectionPointFilteredTray,
        tilesInLayout: TileTransform[],
        randomPoint: Vector
    ): TileTransform {
        for (let i = 0; i < this.populationParams.countOfTriesToInsertTile; ++i) {
            const randomedTile: Tile = pointFilteredTray.randomTile();
            if (randomedTile == null) {
                break;
            }

            const tileTransform: TileTransform = new TileTransform(randomedTile, randomPoint, RandomHelper.nextFloat(0, 360));
            if (
                this.testIfTileMaintaininMarginDistance(
                    tileTransform,
                    new IterableUnion([this.neighbourTiles, tilesInLayout]),
                    this.section.parent.tileMargin
                )
            ) {
                return tileTransform;
            }
        }

        return null;
    }

    private testIfTileMaintaininMarginDistance(pretenderTile: GeometryObject, tilesToTest: Iterable<GeometryObject>, margin: number): boolean {
        for (const tile of tilesToTest) {
            if (!pretenderTile.isMaintainingMinimalDistance(tile, margin)) {
                return false;
            }
        }

        return true;
    }

    private testIfPointIsInsideTile(tilesToTest: Iterable<GeometryObject>, point: Vector): boolean {
        for (const tile of tilesToTest) {
            if (PresenceInPoligonHelper.isPointInsidePolygon(tile.getWorldVertices(), point)) {
                return true;
            }
        }

        return false;
    }

    private getMinDistanceToNeighbourTiles(tilesToTest: IterableUnion<GeometryObject>, point: Vector): number | null {
        let minDistance: number | null = null;
        for (const tile of tilesToTest) {
            const distance: number = PointToPolygonDistanceCalculator.calculatePointToPolygonDistance(point, tile.getWorldVertices());

            if (!minDistance || distance < minDistance) minDistance = distance;
        }

        return minDistance;
    }

    private addTile(layout: PossibleLayout, pointFilteredTray: SectionPointFilteredTray, tileTransform: TileTransform): void {
        pointFilteredTray?.markTileUsed(tileTransform.tile);
        layout.layoutTilesRates.set(tileTransform, this.evaluator.rateTile(tileTransform));
    }
}

File: backend/Irrtirer/Irrtirer.Generator/Orderer/SectionLayout/SectionLayoutPreparer.cs
using Irrtirer.Generator.ColorCompatibility;
using Irrtirer.Generator.Helpers.PolygonHelpers;
using Irrtirer.Generator.Helpers;
using Irrtirer.Generator.Models.GenerationParams;
using Irrtirer.Generator.MosaicStructure;
using Irrtirer.Generator.Tiles;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Numerics;
using static Irrtirer.Generator.Tiles.SectionPointFilteredTray;

namespace Irrtirer.Generator.Orderer.SectionLayout
{
    internal class SectionLayoutPreparer
    {
        private readonly SectionModel section;
        private readonly RandomFactory randomFactory;
        private readonly BlurredImageObject pixelSource;
        private readonly TileTransform[] neighborTiles;
        private readonly PopulationParams populationParams;
        private readonly PossibleLayoutEvaluator evaluator;

        public SectionLayoutPreparer(RandomFactory rnd, BlurredImageObject pixelSource, SectionModel section, TileTransform[] neighborTiles)
        {
            this.randomFactory = rnd;
            this.pixelSource = pixelSource;
            this.section = section;
            this.neighborTiles = neighborTiles;
            this.populationParams = section.Parent.PopulationParams;
            this.evaluator = new PossibleLayoutEvaluator(section.Parent.EvaluationParams, pixelSource, section);
        }

        public PossibleLayout CreateInitialLayout(Tile[] availableTiles)
        {
            PossibleLayout layout = new PossibleLayout(section)
            {
                UsedTilesIndexes = new HashSet<int>(),
                TilesAvailableForLayout = availableTiles
            };

            Random random = randomFactory.GetRandomObject();
            SectionPointFilteredTray pointFilteredTray = new SectionPointFilteredTray(random, availableTiles, layout.UsedTilesIndexes);

            for (int i = 0; i < populationParams.CountOfTrianglePositionDraws && pointFilteredTray.HasAnyAvailableTiles(); ++i)
            {
                TryAddTile(layout, pointFilteredTray, random);
            }

            layout.Rate = evaluator.Evaluate(layout);
            return layout;
        }

        public PossibleLayout ModifyLayout(PossibleLayout layout, int count)
        {
            Random random = randomFactory.GetRandomObject();
            SectionPointFilteredTray pointFilteredTray = new SectionPointFilteredTray(random, layout.TilesAvailableForLayout, layout.UsedTilesIndexes);

            for (int i = 0; i < count; ++i)
            {
                ModificationOperation possibleOperations = ModificationOperation.None;
                if (pointFilteredTray.HasAnyAvailableTiles())
                {
                    possibleOperations |= ModificationOperation.Add;
                }
                if (layout.LayoutTilesRates.Any())
                {
                    possibleOperations |= ModificationOperation.Transform;
                    possibleOperations |= ModificationOperation.Change;
                    possibleOperations |= ModificationOperation.Rotate;
                }

                ModificationOperation operation = random.RandomModificationOperation(possibleOperations);
                switch (operation)
                {
                    case ModificationOperation.Add:
                        TryAddTile(layout, pointFilteredTray, random);
                        break;
                    case ModificationOperation.Change:
                        TryReplaceTile(layout, pointFilteredTray, random);
                        break;
                    case ModificationOperation.Transform:
                        TryMoveTile(layout, pointFilteredTray, random);
                        break;
                    case ModificationOperation.Rotate:
                        TryRotateTile(layout, pointFilteredTray, random);
                        break;
                }
            }

            layout.Rate = evaluator.Evaluate(layout);
            return layout;
        }

        private bool TryAddTile(PossibleLayout layout, SectionPointFilteredTray pointFilteredTray, Random random)
        {
            Vector2 randomPoint = random.RandomPointInTriangle(section.Triangle.Vertices);
            if (!TestIfPointIsInsideTile(neighborTiles.Union(layout.TilesLayout), randomPoint))
            {
                pointFilteredTray.SetFilterForPoint(new SectionPointTrayFilter()
                {
                    MaxInnerRadius = GetMinDistanceToNeighborTiles(neighborTiles.Union(layout.TilesLayout), randomPoint) ?? float.MaxValue,
                    PreferredColor = pixelSource.GetPictureColorAtPosition(randomPoint) ?? Color.Transparent,
                    CountOfColorMatchingAttempts = populationParams.CountOfColorMatchingAttempts
                });

                TileTransform tileTransform = TryFindTileTransformForPoint(pointFilteredTray, layout.TilesLayout, randomPoint);
                if (tileTransform != null)
                {
                    AddTile(layout, pointFilteredTray, tileTransform);
                    return true;
                }
            }

            return false;
        }

        private bool TryMoveTile(PossibleLayout layout, SectionPointFilteredTray pointFilteredTray, Random random)
        {
            TileTransform tile = layout.LayoutTilesRates.ElementAt(random.Next(layout.LayoutTilesRates.Count)).Key;

            for (int i = 0; i < populationParams.CountOfTriesToInsertTile; ++i)
            {
                float newAngle = (tile.Angle + random.Next(-15, 15)) % 360;

                Vector2 newPosition;
                do
                {
                    newPosition = tile.Position + random.NextPointInsideCircle() * section.Parent.TileMinRadius / 100;
                }
                while (!PresenceInPolygonHelper.IsPointInsideTriangle(newPosition, section.Triangle.Vertices));

                TileTransform rotatedTile = tile.Transform(newPosition, newAngle);

                if (TestIfTileMaintainsMarginDistance(rotatedTile, neighborTiles.Union(layout.TilesLayout).Except(new[] { tile }), section.Parent.TileMargin))
                {
                    RemoveTile(layout, pointFilteredTray, tile);
                    AddTile(layout, pointFilteredTray, rotatedTile);
                    return true;
                }
            }

            return false;
        }

        private bool TryRotateTile(PossibleLayout layout, SectionPointFilteredTray pointFilteredTray, Random random)
        {
            TileTransform tile = layout.LayoutTilesRates.ElementAt(random.Next(layout.LayoutTilesRates.Count)).Key;

            for (int i = 0; i < populationParams.CountOfTriesToInsertTile; ++i)
            {
                TileTransform rotatedTile = tile.Transform(tile.Position, random.Next(0, 360));

                if (TestIfTileMaintainsMarginDistance(rotatedTile, neighborTiles.Union(layout.TilesLayout).Except(new[] { tile }), section.Parent.TileMargin))
                {
                    RemoveTile(layout, pointFilteredTray, tile);
                    AddTile(layout, pointFilteredTray, rotatedTile);
                    return true;
                }
            }

            return false;
        }

        private bool TryReplaceTile(PossibleLayout layout, SectionPointFilteredTray pointFilteredTray, Random random)
        {
            TileTransform tile = layout.LayoutTilesRates.ElementAt(random.Next(layout.LayoutTilesRates.Count)).Key;
            IEnumerable<TileTransform> tilesInLayoutWithoutReplaced = layout.TilesLayout.Except(new[] { tile });

            pointFilteredTray.SetFilterForPoint(new SectionPointTrayFilter()
            {
                MaxInnerRadius = GetMinDistanceToNeighborTiles(neighborTiles.Union(tilesInLayoutWithoutReplaced), tile.Position) ?? float.MaxValue,
                PreferredColor = pixelSource.GetPictureColorAtPosition(tile.Position) ?? Color.Transparent,
                CountOfColorMatchingAttempts = populationParams.CountOfColorMatchingAttempts
            });

            TileTransform tileTransform = TryFindTileTransformForPoint(pointFilteredTray, tilesInLayoutWithoutReplaced, tile.Position);
            if (tileTransform != null)
            {
                RemoveTile(layout, pointFilteredTray, tile);
                AddTile(layout, pointFilteredTray, tileTransform);
                return true;
            }

            return false;
        }

        private TileTransform TryFindTileTransformForPoint(SectionPointFilteredTray pointFilteredTray, IEnumerable<TileTransform> tilesInLayout, Vector2 randomPoint)
        {
            for (int i = 0; i < populationParams.CountOfTriesToInsertTile; ++i)
            {
                Tile drawnTile = pointFilteredTray.RandomTile();
                if (drawnTile == null)
                {
                    break;
                }

                TileTransform tileTransform = new TileTransform(drawnTile, randomPoint, randomFactory.Next(0, 360));
                if (TestIfTileMaintainsMarginDistance(tileTransform, neighborTiles.Union(tilesInLayout), section.Parent.TileMargin))
                {
                    return tileTransform;
                }
            }

            return null;
        }

        private bool TestIfTileMaintainsMarginDistance(GeometryObject pretenderTile, IEnumerable<GeometryObject> tilesToTest, float margin)
        {
            foreach (GeometryObject tile in tilesToTest)
            {
                if (!pretenderTile.IsMaintainingMinimalDistance(tile, margin))
                    return false;
            }

            return true;
        }

        private bool TestIfPointIsInsideTile(IEnumerable<GeometryObject> tilesToTest, Vector2 point)
        {
            foreach (GeometryObject tile in tilesToTest)
            {
                if (PresenceInPolygonHelper.IsPointInsidePolygon(tile.GetWorldVertices(), point))
                    return true;
            }

            return false;
        }

        private float? GetMinDistanceToNeighborTiles(IEnumerable<GeometryObject> tilesToTest, Vector2 point)
        {
            float? minDistance = null;
            foreach (GeometryObject tile in tilesToTest)
            {
                float distance = PointToPolygonDistanceCalculator.CalculatePointToPolygonDistance(point, tile.GetWorldVertices());

                if (!minDistance.HasValue || distance < minDistance.Value)
                    minDistance = distance;
            }

            return minDistance;
        }

        private void AddTile(PossibleLayout layout, SectionPointFilteredTray pointFilteredTray, TileTransform tileTransform)
        {
            pointFilteredTray?.MarkTileUsed(tileTransform.Tile);
            layout.LayoutTilesRates[tileTransform] = evaluator.RateTile(tileTransform);
        }

        private void RemoveTile(PossibleLayout layout, SectionPointFilteredTray pointFilteredTray, TileTransform tileTransform)
        {
            pointFilteredTray?.ReleaseTile(tileTransform.Tile);
            layout.LayoutTilesRates.Remove(tileTransform);
        }
    }
}
